### ISSUE TITLE
Emit ES6 modules

### ISSUE THREAD

**Issue by Pauan** [purescript/purescript#2574]
`rollup-plugin-purs` supports both ES6 and CommonJS modules.

Therefore, after `rollup-plugin-purs` replaces `psc-bundle` we can change `psc` to emit ES6 modules in a backwards-compatible way.

**Comment by paf31** [comment]
Sounds good, although without `language-javascript` changes, this would mean we wouldn't be able to check the FFI imports in `psc` itself.

**Comment by hdgarrood** [comment]
Yeah, my impression was that we would only look at this after we've resolved #2545.

**Comment by paf31** [comment]
Right, but the FFI check happens in `psc`, not `psc-bundle`.

**Comment by hdgarrood** [comment]
Oh ok, I was assuming that resolving #2545 would involve deciding what to do about that check - whether we can afford to remove it from `psc` and leave it to a bundling tool to detect those errors, or contributing to or forking `language-javascript` so that it can deal with ES6 modules, or something else entirely.

**Comment by paf31** [comment]
I think I'd be okay with leaving that to an external bundling tool, at least for 1.0.

**Comment by paf31** [comment]
I assume we don't want to break this in 0.11? This would obviously break just about every package out there, unless we can find a way to support both types of module.

**Comment by hdgarrood** [comment]
I think we first need to come up with a way of doing DCE at a similar quality to `psc-bundle` on ES6 modules. I think Rollup is currently too conservative, see my comments on #2545.

**Comment by Pauan** [comment]
@paf31 We shouldn't do this until after `rollup-plugin-purs` becomes the de facto way of bundling. After `rollup-plugin-purs` becomes popular, this won't even be a breaking change.

@hdgarrood In my opinion both of those issues are quite minor in practice, and even with those issues Rollup *still* manages to have dramatically smaller output compared to `psc-bundle`, so I'm not worried.

And any further optimizations to Rollup can be added as a non-breaking change in the future.

And if Rollup can't achieve the same level of DCE as `psc-bundle`, then I'll just modify `rollup-plugin-purs` to perform DCE. Since PureScript is so pure, it shouldn't be that hard to add.

**Comment by hdgarrood** [comment]
The way I understand it, Rollup is currently unable to remove unused data constructors and unused curried functions; since a rather large proportion of PureScript code falls into one of those two categories, it appears to me to be quite a serious problem.

**Comment by hdgarrood** [comment]
Also, in https://github.com/Pauan/purescript-tree-shaking, rollup only beats psc-bundle if you're looking at the size of the JS after running uglify, and judging by what is printed to stdout, it seems like that's because uglify is doing a lot of additional DCE too (lots of lines of the form "Dropping unused variable foo"). I'd rather not rely on having two separate tools to get a similar quality of DCE.

**Comment by Pauan** [comment]
@hdgarrood The problem with unused data constructors is completely solved by removing the IIFE from PureScript's output. @garyb has said that this is a trivial change to make in `psc`

Removing unused curried functions is a very trivial change to make in `rollup-plugin-purs`

You do realize that UglifyJS prints **far** more warnings about unused code for `psc-bundle` compared to Rollup, right? Rollup prints 27 warnings, `psc-bundle` prints 274!

There's no point in even comparing unminified numbers. If you care even a tiny bit about size, you'll minify in production. This applies not just to PureScript, but to JavaScript as well.

I'm genuinely confused as to why you're worried. These are not big problems. They are easily solved. Some of the problems are already being solved as we speak.

**Comment by hdgarrood** [comment]
> The problem you found was about a curried function being used at the top-level, which as far as I know only happens once with unsafeCompare. It is not a widespread problem, it's a very specific bug, which is easily fixed.

I don't understand why this is not a widespread problem; won't any top-level declaration in PureScript become a top-level declaration in the JS? I'm glad to hear it's easily fixed though.

> You do realize that UglifyJS prints more warnings about unused code for psc-bundle compared to Rollup, right? Rollup prints 158 warnings, psc-bundle prints 274.

On my machine, after redirecting the build logs to files:

```
$ grep 'Dropping \(unused\|unreachable\)' rollup.log | wc -l
133
$ grep 'Dropping \(unused\|unreachable\)' psc-bundle.log | wc -l
4
```

From my reading of what the other warnings say, they're not indications that uglify is eliminating unused code.

> And most of the warnings for Rollup are not because of Rollup's DCE, instead it is because of the uncurrying optimization (notice how most of the warnings have an _uncurried at the end). With uncurrying and inlining disabled, it only prints 27 warnings. And as we speak I'm working on a fix for that issue.

Ok cool, that makes me happier. :smile: 

> There's no point in comparing unminified numbers. If you care even a tiny bit about size, you'll minify in production. This applies not just to PureScript, but to JavaScript as well.

I understand this, but I don't want to have to assume that people are using uglifyjs (or some other minifier that is as effective at removing dead code). I think it's likely that some people are using `hjsmin` on PureScript output, for instance, and I don't think `hjsmin` tries to identify and remove unused code. I also want 

> These are not big problems. They are easily solved.

I really do hope you're right, but in my experience it's not all that uncommon for me to consider some problem to be easily solved right up until I actually try to solve it. Don't get me wrong, I do really appreciate all the work you've done and I would like to see us switch to ES6 modules, but I would prefer to solve these problems before we start switching over.

**Comment by Pauan** [comment]
@hdgarrood No, it only applies to a situation like this:

```
var foo = bar(1)(2)(3);
```

In other words, *calling* a curried function at the top level. *Defining* a curried function at the top level should be fine.

Oddly enough, I just checked again, and it seems that `unsafeCompare` is actually being used, so it *shouldn't* be tree-shaken. Bizarre.

In any case, I plan to implement a custom tree shaker for `rollup-plugin-purs` that should do a better job of removing unused code compared to Rollup (because it can be specialized for PureScript).

Even in the absolute worst case scenario, `rollup-plugin-purs` can just use whatever tree shaking algorithm `psc-bundle` is using, so it shouldn't be any *worse* than `psc-bundle`.

> From my reading of what the other warnings say, they're not indications that uglify is eliminating unused code.

The other `Side effects in initialization of unused variable` warnings are also about dead code. Most of them are about imports which aren't used, since `psc-bundle` doesn't seem to prune a `var Foo = PS["Foo"]` statement even if it's not used.

> I think it's likely that some people are using hjsmin on PureScript output, for instance, and I don't think hjsmin tries to identify and remove unused code.

That's fair. It's not a bad idea to try other minifiers, including weak ones like JSMin.

> I really do hope you're right, but in my experience it's not all that uncommon for me to consider some problem to be easily solved right up until I actually try to solve it.

I know that feeling. The reason I'm not worried is because `rollup-plugin-purs` is already doing **very** good, even without much work put into it, and I know what Babel and Rollup's plugin system is capable of, so I'm confident the existing problems can be solved. As for whether any *new* problems show up, well, that's a different story.

> Don't get me wrong, I do really appreciate all the work you've done and I would like to see us switch to ES6 modules, but I would prefer to solve these problems before we start switching over.

Thank you. But I think the real blocker for this issue is whether `rollup-plugin-purs` becomes popular or not, and that's going to take a while, regardless of whether we fix those bugs or not.

In any case, I can work on the tree shaking in `rollup-plugin-purs`, but somebody else will need to fix the problem with IIFE in `psc`, because I don't know my way around the PureScript codebase.

**Comment by hdgarrood** [comment]
> In other words, calling a curried function at the top level. Defining a curried function at the top level should be fine.

Right, that makes sense. I think I would still expect to see functions to be called at the top level fairly often, though.

> Oddly enough, I just checked again, and it seems that unsafeCompare is actually being used, so it shouldn't be tree-shaken. Bizarre.

I was testing with a project that I had just created with `pulp init`, i.e. where `main` is nothing more than `log "Hello, sailor!"`, so I was expecting everything to be removed apart from `main` and the definition of `log` from `Control.Monad.Eff.Console`.

Regarding adoption of `rollup-plugin-purs`, I'm not too worried, since I expect to be able to change `pulp` so that `pulp build --to bundle.js` uses Rollup under the hood instead of psc-bundle. I also expect that people who are not using Pulp should be able to get whichever tools they're using to deal with ES6 modules fairly easily; AFAIK, the two most commonly used tools for building PureScript code are Pulp and Webpack.

**Comment by Pauan** [comment]
@hdgarrood I expected that too, which is why I was surprised that it only happened with `unsafeCompare`. I suspect it had to do with some combination of currying and data constructors (because `unsafeCompare` uses the `LT`, `EQ`, and `GT` data constructors).

Having said that, I took a look at the output from `rollup-plugin-purs@1.0.13`, and I see a lot of other unused functions, like `isNothing`, `isLeft`, `isRight`, `updateAt`, `uncons`, etc.

I'm not sure why I remember `unsafeCompare` being the only unused curried function call. There's been a lot of changes to `rollup-plugin-purs` and `purescript-tree-shaking`, so maybe that changed something, or maybe the programs I was testing were too small.

----

In any case, I just released version `1.0.16` of `rollup-plugin-purs`, which has much better DCE, better uncurrying, and it now also inlines functions too.

The problem with curried function calls not being removed has been completed solved.

With the new version, there is now only 19 warnings from UglifyJS. Here are the new numbers:

Bundler | Unminified | JSMin | UglifyJS
--- | --- | --- | ---
`rollup-plugin-purs` | `117.9 kB` | `78.5 kB` | `41.8 kB`
`psc-bundle` | `165.1 kB` | `112.0 kB` | `69.3 kB`
`webpack` | `580.1 kB` | `418.2 kB` | `244.4 kB`

I think `rollup-plugin-purs` is now competitive with `psc-bundle` for DCE, so the only issue left is the IIFE with data constructors in `psc`

**Comment by hdgarrood** [comment]
Nice! I'll start looking more at using rollup in pulp soon.

**Comment by paf31** [comment]
I think we should discuss pros and cons of implementing ES6 codegen in 1.0 (and then committing to it for the long term).

Pros:

- ES6 features we can benefit from: proper module system, proper lexical scoping via `let` (which will help with TCO codegen), possibility of proper tail calls later.
- ES6 modules are more principled than CommonJS.

Cons:

- Targeting ES3/ES5 obviously means compiled code is usable with a wider variety of targets, whereas using ES6 in the browser basically means using another transpiler like Babel _after_ `psc`. There are plenty of users who won't/can't use PureScript if it means running Node tools as well as `psc`. 
- ES6 modules are less hackable than CommonJS. PSCi in the browser won't be able to work without writing some sort of ES6 parser, whether it's a regex hack or a full blown parser implementation.

What did I miss? Let's please try to come to a conclusion on this relatively soon, since it's the largest remaining item for 1.0.

**Comment by rvion** [comment]
As of now, we don't need any post processing or bundling when targeting node: we'll lose that.

Since PureScript aims to Target multiple languages, can't we have both commonjs __and__ es6 code generation, with some flag switch?

**Comment by paf31** [comment]
> Since PureScript aims to Target multiple languages, can't we have both commonjs and es6 code generation, with some flag switch?

Yes, I've thought about that option too. It's good from the point of view of things like PSCi, and will  tackle the major problem with ES3/ES6 (usability without Babel), but it does mean maintaining more code going forward. We might want to put some effort into sharing code between the two backends if we go that route (TCO, for example, will not be exactly the same, but the part of the code which identifies what can be optimized should be shared).

**Comment by garyb** [comment]
Perhaps that can be something expressable in "vague" terms in `CoreImp`, and is transformed into the approprate code in an AST for ES3/ES6.

**Comment by garyb** [comment]
Err, the TCO that is, FFI is still a problem.

**Comment by paf31** [comment]
One unpleasant option is to provide both ES3 and ES6 FFI modules.

**Comment by paf31** [comment]
Another option is to stick with ES3/5, and to improve CoreImp to the point where an ES6 backend could be built separately, with its own package sets.

**Comment by bodil** [comment]
I could see the advantage to being able to emit both CJS and ES6 modules, but that doesn't imply needing to support ES3 - every modern browser implements more or less all of ES6 apart from modules, and so does Node. If anyone still desperately needs IE6 compatibility, they can use Babel.

**Comment by Cmdv** [comment]
If people aren't aware of some of the great features that come with ES6 then take a look [HERE](https://github.com/lukehoban/es6features).

The modules, exports could come in real handy when dealing with multiple thing the strongest I can think of is the use of DCE.

Rest and spread operators as well as destructuring, the arrow functions and their lexical scoping.
I'm sure these things could greatly improve the output from purescript as well as size.

Also if you want to keep ES5 then you can use tools to convert JS to ES5 😄

**Comment by paf31** [comment]
> very modern browser implements more or less all of ES6

True, but there are more targets than just browsers and Node. Do we know that all of the ones people are interested in using support the same features?

Modules were the main reason I wanted to move to ES6, but they are also the most troublesome from a maintenance point of view. It is conceivable that we could target ES6-minus-modules though, but it seems a bit messy to me.

**Comment by rightfold** [comment]
ES3 works everywhere ES6 works. Not the other way around. It'd wait a decade.

**Comment by spicydonuts** [comment]
ES6 (and beyond).  Compiling ES6 to ES[5,3] already has a good story when doing so is necessary.  It could even be part of a tool like `pulp`.  The reverse is _not_ true.  ES6 -> ES3 _loses information_.  This most strongly affects bundling and code elimination.  If ES6 is not an option for some reason then the compiler target should be Closure so that information is not lost.

**Comment by paf31** [comment]
> It could even be part of a tool like pulp

Yes, but that makes things harder for users who don't use Pulp. Users who don't use Node tooling are left with no options if they want to compile away ES6 modules.

All of the arguments for ES6 I've heard here can be answered with an alternative CoreFn backend and package set. Any thoughts on that?

My main reason for wanting ES6 was so that we had a proper module system with which to do dead code elimination. Now that dead code elimination is not in scope for 1.0, I'm less worried about that.

**Comment by Cmdv** [comment]
> My main reason for wanting ES6 was so that we had a proper module system with which to do dead code elimination. Now that dead code elimination is not in scope for 1.0, I'm less worried about that.

If the output is ES6 it will also open up to the use of lots of Javascript tools not just DCE

**Comment by garyb** [comment]
Not meaning to derail, but

> My main reason for wanting ES6 was so that we had a proper module system with which to do dead code elimination. Now that dead code elimination is not in scope for 1.0, I'm less worried about that.

When did this happen? DCE seems like rather a big thing to drop...

**Comment by paf31** [comment]
@garyb It's [here](https://github.com/purescript/purescript/issues/2545). I'm not suggesting dropping `psc-bundle` completely, but reducing the scope of what we support in the compiler distribution itself. The DCE code can be supported post 1.0 outside the compiler distro, much like `psc-package`, until we have a better solution.

**Comment by sudhirvkumar** [comment]
I am not a fan of ES6, mainly because I would like to avoid using `babel`. The main reason for me to look at PureScript is to avoid `The JavaScript Problem` along with `node_modules` problem and `npm` problem.

Also, Early release of PureScript 1.0 is very important for user perception on Language stability and maturity. That will enable many people to sell PureScript to their team!

Working on a fork of `language-javascript` isn't something which is going to help speed up the release of 1.0 and  FFI import checking is critical.

TCO, we can always improve that in 1.1 or later?

DCE, we can always use external tools to achieve that if required!

For JavaScript code quality, may be we can force a config of `eslint` on all packages and show those warning(s) in the `psc-package` and `pursuit` so that package maintainers can follow the same coding style and quality checks for FFI code, there by making PureScript Applications more robust & reliable.

**Comment by Pauan** [comment]
@sudhirvkumar [`rollup-plugin-purs`](https://github.com/pauan/rollup-plugin-purs) does do import checking (and a lot of other stuff too).

Right now it seems like `rollup-plugin-purs` is going to replace `psc-bundle` for most use-cases, which is why @paf31 is considering reducing the features from `psc-bundle`

**Comment by sudhirvkumar** [comment]
>  `rollup-plugin-purs` does do import checking (and a lot of other stuff too).

@Pauan thats interesting. I think FFI imports should be checked by `purs` and not just at the bundling stage. When compiler tests then it helps resolve things during the development stage and checking during bundling is too late... we normally bundle when we are ready to go live.

Actually we should do more FFI quality checks! like `eslint`... can `rollup-plugin-purs` help in that area?

Thanks for all the hard work and effort on `rollup-plugin-purs` we will be using it more in the future.

**Comment by Pauan** [comment]
@sudhirvkumar Are you creating a Node app with PureScript? If doing browser development, bundling is mandatory anyways, so you'll get import warnings at development time.

And when using `watch` mode, recompilation is pretty fast (but not as fast as I would like), so you can use it in development.

I agree that something like `eslint` would be great. It seems somebody already created [some rules which validate ES6 modules](https://github.com/benmosher/eslint-plugin-import).

I think it would be cool if the default PureScript package set ran `eslint` on all packages, but that's a separate issue from ES6 modules.

I think `eslint` is outside of the scope of `rollup-plugin-purs`, but you can set up your `eslint` configuration to use the `output` directory. Or you can use [`rollup-plugin-eslint`](https://github.com/TrySound/rollup-plugin-eslint)

**Comment by Cmdv** [comment]
@paf31 
> we should aim to get the functional/imperative core into a state where it is possible to develop ES6 support (with its own FFI) as a separate backend without too much effort.

can you point me to an example to a backend, I'm really keen to try help out support ES6?

**Comment by garyb** [comment]
@Cmdv the current JS backend is all under here: https://github.com/purescript/purescript/tree/master/src/Language/PureScript/CodeGen

**Comment by starper** [comment]
@paf31 sorry, but I don't quite get it, as far as I understand, you are ok with reducing functionality of `purs bundle` in favour of using `Rollup` (it's a separate tool) for DCE, but at the same time you are somehow sceptical about using `Babel` (wich is a separate tool too and and can be used as a plugin as well) for ES6.

Btw +1 for ES6.

**Comment by robotlolita** [comment]
A thing to consider before moving to ES6 modules is that Node still hasn't decided how it'll implement it, and they're talking to TC39 to get the ES spec changed in some places https://github.com/nodejs/node-eps/blob/master/002-es6-modules.md#5-semantics. This is likely not going to result in any big changes, but it might result in some small backwards-incompatible changes in Node at least (as in, previously compiled modules would have to be recompiled with new rules).

**Comment by sudhirvkumar** [comment]
@starper reducing functionality of `purs bundle` enables better flexibility of toolchain in the future. `rollup` and DCE is optional. If we go with ES6 we need `babel` by default. You can't work with PureScript unless you are ready to configure and work with babel!

**Comment by starper** [comment]
@sudhirvkumar that's exactly the argumentation that I don't quite get. Current support of ES6 (except for `modules`) in modern browsers and latest `Node.js` makes `Babel` or any other transpiler as optional as `Rollup`.

Anyway, I'm ok with having ES6 as a separate backend. The only thing that worries me is the possibility that, since PureScript is evolving very fast, ES6 backend may get much less attention then main ES3/5 backend.

**Comment by bmeck** [comment]
Node.js won't be shipping ESM til next year LTS. LMK if you have any questions about things. Biggest differences: URLs are used for resolution, magic CJS variables (__dirname, __filename, require) are not in ESM bootstrap,  and no named imports from CJS.

**Comment by Pauan** [comment]
Here are some additional facts:

* ES6 modules are the future. So it's not a choice between CommonJS and ES6, it's a choice between ES6 now and ES6 later. If we don't support ES6 modules in 1.0, we'll need to add them in a future breaking version (like 2.0, 3.0, etc.)

* A lot of other compile-to-JavaScript languages have support for ES6 modules, or are currently working on adding in support. This is starting to become expected behavior (because of the previous fact).

* Rollup and Webpack can support ES6 and CommonJS simultaneously in the same program, easing the transition to ES6. Despite that, the transition will still be rough. It is a major breaking change after all.

* Bundlers are already used for web development, so the people who are primarily affected by the move to ES6 modules are people writing Node.js apps in PureScript.

* There are two separate issues: whether `psc` emits ES6 modules, and whether FFI files use ES6 modules. The first issue is a breaking change that mostly affects Node.js apps and PureScript tools. The second issue is a breaking change that affects everybody, since many libraries will need to be changed.

* Supporting ES6 modules does **not** mean we need to use Babel. Rollup and Webpack both have built-in support for ES6 modules.

* If `psc` emits ES6 features which are [commonly supported](https://kangax.github.io/compat-table/es6/), and avoids ES6 features which are not commonly supported, then we can take advantage of (some) ES6 features without needing Babel. Over time the number of ES6 features we can use will increase.

* If Rollup/Webpack are too slow for compilation (during development), it shouldn't be hard to write a custom ES6 bundler which doesn't do any tree shaking, so it can cache modules and run much faster.

And here are my thoughts:

* I do like ES6 modules, and I think it would be nice to have them supported in 1.0, but I also don't mind waiting for native support in Node and browsers. Just keep in mind that if they don't get added to 1.0, they will need to be implemented eventually.

* I think if we're going to start emitting ES6 modules from `psc`, we should also move FFI files over to ES6 modules at the same time. If changing the FFI files is too big of a change for 1.0, then we'll just have to wait until 2.0

**Comment by Pauan** [comment]
@bmeck Great, thanks for the help! I really appreciate the work you've done on ES6 modules in Node.

* What do you mean by "URls are used for resolution"? I was under the impression that the module specifiers would be the same as with CommonJS.

* If you need to access `__dirname`, `__filename`, etc. I assume that will be done with a built-in module, similar to [this proposal](https://github.com/whatwg/html/issues/1013#issuecomment-259770765). Is that correct?

* The last time I read the proposal, it [said that named exports are supported](https://github.com/nodejs/node-eps/blob/master/002-es6-modules.md#54-es-consuming-commonjs). Do you have a link to the most up-to-date version of the proposal?

**Comment by bmeck** [comment]
@Pauan 

> What do you mean by "URls are used for resolution"? I was under the impression that the module specifiers would be the same as with CommonJS.

Still slightly under discussion ( https://github.com/nodejs/node-eps/issues/51 ). Mostly benign, but observable differences since URL schemes are parsed / query and hash differences load different modules / URL encoding means some odd filename decoding is different.

> I assume that will be done with a built-in module, similar to this proposal. Is that correct?

Yup.

> Do you have a link to the most up-to-date version of the proposal?

https://github.com/nodejs/node-eps/pull/39 , yet to be merged as implementation makes minor tweaks to it.

**Comment by paf31** [comment]
> ES6 modules are the future.

@Pauan Do you see the possibility for browsers to implement ES6 modules fully, any time soon?

I'm all in favor of switching to ES6 as the default in 2.0, but I would want to wait for browser support. Many of our users come from Haskell, and don't want to use Node tools (be it Rollup or Babel or something else) for whatever reason.

**Comment by Pauan** [comment]
@bmeck Thanks!

----

@paf31 v8 already [fully supports ES6 modules since January](https://bugs.chromium.org/p/v8/issues/detail?id=1569#c153), and the Chromium team is [doing steady work at implementing `<script type="module">`](https://bugs.chromium.org/p/chromium/issues/detail?id=594639)

According to @bmeck we can expect ES6 modules in Node within 1-2 years. I don't know how far along the non-Chrome browsers are, but 1-2 years sounds about right to me.

**Comment by Pauan** [comment]
@paf31 Also, I fully understand the desire to avoid weird JavaScripty tools, but most PureScript users use Pulp, and Pulp provides an easy-to-use wrapper around the JavaScripty tools.

So personally, I think it's acceptable to tell people to "just do `pulp build`". Then they don't need to think about or care about what Pulp uses under the hood.

This applies to both Haskellers and non-Haskellers. I'm a heavy JavaScript programmer, and I have a lot of knowledge about JavaScript tools, but I appreciate the simplicity of Pulp, and I use it regularly.

As for the people who don't use Pulp, I assume they're knowledgeable enough to know how to use whatever tool they're using, whether that be Rollup, purs-loader, or whatever.

**Comment by michaelficarra** [comment]
@paf31 @Pauan `<script type=module>` has been implemented in Webkit and made available in Safari since Tech Preview 21, back in early January. And other implementations are not far behind. Support is not years away.

@Pauan I wanted to say I agree with everything in [your earlier comment](https://github.com/purescript/purescript/issues/2574#issuecomment-291511462). There's a very strong argument for using ECMAScript modules in PureScript 1.0.

**Comment by Pauan** [comment]
@michaelficarra That's great news! I just read [this article](https://www.contentful.com/blog/2017/04/04/es6-modules-support-lands-in-browsers-is-it-time-to-rethink-bundling/) which suggests that Firefox and Edge will have support very soon as well.

So it seems Node is the biggest impediment, especially because people writing Node.js apps in PureScript are used to not bundling.

I don't think it's an unsolvable problem, however. Especially if Pulp gets built-in support for Rollup, we just have to tell people to run `pulp build --to foo.js` (or `pulp rollup` or whatever) even on Node. It's not as nice of a workflow as before, but it's doable.

**Comment by Pauan** [comment]
@michaelficarra @paf31 Also, keep in mind that the browsers are currently releasing ES6 modules on their unstable branches, often behind a flag. That means it'll still be a while before ES6 modules are supported in stable versions. At least a few months, but probably less than a year.

**Comment by paf31** [comment]
I appreciate that Pulp hides a lot of the complexity, and I use Pulp myself for that reason. but it's important to me that PureScript doesn't require Pulp in order to provide basic features. I think the first step should be to prototype an ES6 backend as a fork of the compiler. That way, we can test it out, build a package set, etc. Until then, the compiler can continue to target the widest array of browser and non-browser targets.

**Comment by Pauan** [comment]
@paf31 I like the idea of changing `psc-bundle` so that it understands ES6 modules, either by using an already-existing ES6 bundler, or creating a custom ES6 bundler.

We need `psc-bundle` anyways, because of PSCi and Try PureScript. And this means that there is a built-in way of bundling ES6 modules, without any reliance on third-party tools.

Because it doesn't do any tree shaking or optimizations, it can be very fast, and it can cache modules for even more speed. This speed is useful for people writing Node.js apps, but it's also very useful during development (`rollup-plugin-purs` emits very small code, but it's very slow)

And because it only handles ES6 modules, it can produce fast, lightweight code (no need for a `require` shim).

So `pulp build` would use the new `psc-bundle` (very fast, great for development/Node apps), and `pulp build --optimize` would use `rollup-plugin-purs` (very small file size, great for production)

Of course people could also use `psc-bundle` by itself, without any reliance on pulp.

**Comment by kevinbarabash** [comment]
Now that a number of browsers support ES6 modules, one benefit of emitting ES6 modules is that dev builds become super fast because you wouldn't have to bundle the files.  I suppose a person could shim `require` in the browser for the same purpose.

**Comment by hdgarrood** [comment]
Closing in favour of #3613.
